File: prohash/javascript/prohash.js
const ffi = require('ffi-napi');
const ref = require('ref-napi');
const StructType = require('ref-struct-napi');

/**
 * ProHash-256 Node.js Wrapper ("The Shredder")
 * Benötigt: ffi-napi, ref-napi, ref-struct-napi
 */

// --- Typ-Definitionen ---
const uint32 = ref.types.uint32;
const uint8 = ref.types.uint8;
const uint64 = ref.types.uint64;
const size_t = ref.types.size_t;

// Definition des ProHash_Ctx (136 Bytes)
const ProHashCtx = StructType({
    'belt': ffi.Array(uint32, 16),   // 512-Bit Interner Zustand
    'count': uint64,                // Anzahl verarbeiteter Bytes
    'buffer': ffi.Array(uint8, 64),  // Eingabe-Puffer
    'buf_idx': uint32               // Aktuelle Position
});

// Pfad zur Library (automatische Endung je nach OS)
const libPath = process.platform === 'win32' ? 'prohash.dll' : './libprohash.so';

const lib = ffi.Library(libPath, {
    'prohash_init': ['void', [ref.refType(ProHashCtx)]],
    'prohash_update': ['void', [ref.refType(ProHashCtx), 'pointer', size_t]],
    'prohash_final': ['void', [ref.refType(ProHashCtx), 'pointer']],
    'prohash_get_version_info': ['string', []]
});

class ProHash {
    constructor() {
        this.ctx = new ProHashCtx();
        lib.prohash_init(this.ctx.ref());
        this.finalized = false;
    }

    /**
     * Gibt die Versionsinformationen der nativen Library zurück.
     */
    static getVersion() {
        return lib.prohash_get_version_info();
    }

    /**
     * Verarbeitet Daten (Buffer oder String).
     */
    update(data) {
        if (this.finalized) throw new Error("Hash already finalized.");
        
        const input = Buffer.isBuffer(data) ? data : Buffer.from(data);
        if (input.length === 0) return;

        lib.prohash_update(this.ctx.ref(), input, input.length);
    }

    /**
     * Schließt die Berechnung ab und gibt den 32-Byte Digest als Buffer zurück.
     */
    finalize() {
        if (this.finalized) throw new Error("Hash already finalized.");
        
        const digest = Buffer.alloc(32);
        lib.prohash_final(this.ctx.ref(), digest);
        this.finalized = true;
        
        // Zero-Trust: Kontext im Speicher nullen
        return digest;
    }

    /**
     * Statische Utility für Einmal-Hashing.
     */
    static hash(data, encoding = 'hex') {
        const hasher = new ProHash();
        hasher.update(data);
        const digest = hasher.finalize();
        return digest.toString(encoding).toUpperCase();
    }
}

module.exports = ProHash;

File: prohash/java/ProHash.java
package com.brainai.prohash;

import com.sun.jna.*;
import com.sun.jna.ptr.PointerByReference;
import java.util.Arrays;
import java.util.List;

/**
 * Java Wrapper für den ProHash-256 Algorithmus ("The Shredder").
 * Nutzt JNA zur Anbindung der nativen Core-Bibliothek.
 */
public class ProHash {

    private static final String LIB_NAME = "prohash";

    // --- Native Interface Definition ---
    public interface IProHash extends Library {
        IProHash INSTANCE = Native.load(LIB_NAME, IProHash.class);

        void prohash_init(ProHashCtx ctx);
        void prohash_update(ProHashCtx ctx, byte[] data, NativeSize len);
        void prohash_final(ProHashCtx ctx, byte[] digest);
        String prohash_get_version_info();
    }

    /**
     * Mapping der nativen ProHash_Ctx Struktur (136 Bytes).
     * Das Alignment ist auf 8-Byte gesetzt, um dem uint64_t count gerecht zu werden.
     */
    @Structure.FieldOrder({"belt", "count", "buffer", "bufIdx"})
    public static class ProHashCtx extends Structure {
        public int[] belt = new int[16];      // uint32_t belt[16] (512-bit internal state)
        public long count;                    // uint64_t count (Processed bytes)
        public byte[] buffer = new byte[64];  // uint8_t buffer[64] (Input buffer)
        public int bufIdx;                    // uint32_t buf_idx (Buffer position)

        public ProHashCtx() {
            super();
            setAlignType(Structure.ALIGN_DEFAULT);
        }

        @Override
        protected List<String> getFieldOrder() {
            return Arrays.asList("belt", "count", "buffer", "bufIdx");
        }
    }

    private final ProHashCtx ctx;
    private boolean finalized = false;

    public ProHash() {
        this.ctx = new ProHashCtx();
        IProHash.INSTANCE.prohash_init(this.ctx);
    }

    /**
     * Gibt die Versionsinformationen der nativen Library zurück.
     */
    public static String getVersion() {
        return IProHash.INSTANCE.prohash_get_version_info();
    }

    /**
     * Verarbeitet einen Datenblock.
     */
    public void update(byte[] data) {
        if (data == null || data.length == 0) return;
        if (finalized) throw new IllegalStateException("Hash already finalized.");
        
        IProHash.INSTANCE.prohash_update(this.ctx, data, new NativeSize(data.length));
    }

    /**
     * Schließt die Berechnung ab und gibt den 32-Byte Digest zurück.
     * Der interne Kontext wird danach sicher gelöscht.
     */
    public byte[] finalizeHash() {
        if (finalized) throw new IllegalStateException("Hash already finalized.");
        
        byte[] digest = new byte[32];
        IProHash.INSTANCE.prohash_final(this.ctx, digest);
        finalized = true;
        return digest;
    }

    /**
     * Hilfsmethode zur schnellen Berechnung eines Hex-Hashes für Byte-Arrays.
     */
    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}